File: cppversion/30. substring concatenation.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

std::vector<int> findSubstring(std::string s, std::vector<std::string> &words)
{
    std::unordered_map<std::string, int> counts;
    for (std::string word : words)
        counts[word]++;
    int len_str = s.length(), num_of_words = words.size(), len_of_each_word = words[0].length();
    std::vector<int> indices;
    for (int i = 0; i < len_str - num_of_words * len_of_each_word + 1; i++)
    {
        std::unordered_map<std::string, int> seen;
        int j = 0;
        for (; j < num_of_words; j++)
        {
            std::string word = s.substr(i + j * len_of_each_word, len_of_each_word);
            if (counts.find(word) != counts.end()) // if word is found
            {
                seen[word]++;
                if (seen[word] > counts[word])
                    break;
            }
            else
                break;
        }
        if (j == num_of_words)
            indices.push_back(i);
    }
    return indices;
}

int main()
{
    std::string input_str = "barfoothefoobarman";
    std::vector<std::string> words = {"foo", "bar"};
    std::vector<int> indices;
    indices = findSubstring(input_str, words);
    for (int index : indices)
        std::cout << index << " ";
    std::cin.get();
}

File: pyversion/30. substring concatenation.py
from collections import Counter


def findSubstring(s, words):
    counts = Counter()
    for word in words:
        counts[word] += 1
    len_str = len(s)
    num_of_words = len(words)
    len_of_each_word = len(words[0])
    indices = []
    for i in range(len_str - num_of_words * len_of_each_word + 1):
        seen = Counter()
        for j in range(num_of_words):
            word = s[i + j * len_of_each_word:i + (j + 1) * len_of_each_word]
            if word in words:
                seen[word] += 1
                if seen[word] > counts[word]:
                    break
            else:
                break
        if (j + 1) == num_of_words:
            indices.append(i)
    return indices


if __name__ == '__main__':
    print(findSubstring('barfoothefoobarman', ['foo', 'bar']))
